File: 백준/단계 별로 풀어보기/동적_계획법_1/11054_가장_긴_바이토닉_부분_수열.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <tuple>
using namespace std;

int bisect_left(vector<int> &list, int &finding){
	int S = 0, M = 0, E = list.size() - 1;
	        
    while (S < E){
        M = (S + E) / 2;
        if (finding == list[M]) break;
        else if (finding < list[M]) E = M - 1;
        else S = M + 1;
    }
    
    if (list[M] != finding){
    	if (list[S] < finding) return S + 1;
    	else return S;
    }
    return M;
}


int main()
{
	cin.tie(NULL);
	cout.tie(NULL);
	ios_base::sync_with_stdio(false);
	
	int N;
	cin >> N;
	vector<int> data(N);
	for (int i = 0; i < N; i++) cin >> data[i];
	
	vector<int> min_list1(1, data[0]);
	vector<int> min_list2(1, data[N - 1]);
	vector<int> dp1;
	vector<int> dp2;
	int MAX = 0;
	
	for (auto &present : data){
	    if (min_list1[min_list1.size() - 1] < present){
	        min_list1.push_back(present);
	        dp1.push_back(min_list1.size());
	    }
	    else{
	        int index = bisect_left(min_list1, present);
	        min_list1[index] = present;
	        dp1.push_back(index + 1);
	    }
	}
	
	for (int i = N - 1; 0 <= i; i--){
		int &present = data[i];
	    if (min_list2[min_list2.size() - 1] < present){
	        min_list2.push_back(present);
	        dp2.push_back(min_list2.size());
	    }
	    else{
	        int index = bisect_left(min_list2, present);
	        min_list2[index] = present;
	        dp2.push_back(index + 1);
	    }
	}
	
	
	for (int i = 0; i < dp1.size(); i++){
		if (MAX < dp1[i] + dp2[N - i - 1]) MAX = dp1[i] + dp2[N - i - 1];
	}
	
	cout << MAX - 1;
}

File: baekjoon/단계 별로 풀어보기/동적_계획법_1/11054_가장_긴_바이토닉_부분_수열.py
from bisect import bisect_left
import sys
input = sys.stdin.readline
IP = lambda: map(int, input().split())

def solve():
    N = int(input())
    data = list(IP())
    min_list1 = [data[0]]
    min_list2 = [data[-1]]
    dp1 = []; dp2 = []; MAX = 0
    
    for present in data:
        if min_list1[-1] < present:
            min_list1.append(present)
            dp1.append(len(min_list1))
        
        else:
            index = bisect_left(min_list1, present)
            min_list1[index] = present
            dp1.append(index + 1)
    
    for present in data[::-1]:
        if min_list2[-1] < present:
            min_list2.append(present)
            dp2.append(len(min_list2))
            
        else:
            index = bisect_left(min_list2, present)
            min_list2[index] = present
            dp2.append(index + 1)
            
    
    for i, j in zip(dp1, dp2[::-1]):
        if MAX < i + j: MAX = i + j
    
    print(MAX - 1)
    
    
if "__main__" == __name__: solve()